working/0731 - horizontal inference/feature-specification.md

```diff
@@ -0,0 +1,290 @@
+# Horizontal inference
+
+Author: paulberry@google.com, Version: 1.0 (See [Changelog](#Changelog) at end)
+
+Horizontal inference allows certain arguments of an invocation (specifically
+function literals) to be type inferred in a context derived from the type of
+other arguments.
+
+## Motivation
+
+The key motivating example is the Iterable.fold method, which has the following
+signature:
+
+```dart
+abstract class Iterable<E> {
+  T fold<T>(T initialValue, T combine(T previousValue, E element)) { ... }
+}
+```
+
+A typical usage of `fold` today looks like this:
+
+```dart
+Iterable<int> values = ...;
+int largestValue = values.fold(0, (a, b) => a < b ? b : a);
+```
+
+Note the use of the explicit type `int` for `largestValue`.  We would like for
+the user to be able to drop this explicit type, and instead write:
+
+```dart
+Iterable<int> values = ...;
+var largestValue = values.fold(0, (a, b) => a < b ? b : a);
+```
+
+Today this doesn't work, because without the leading `int`, the downwards
+inference phase of type inference has no information with which to choose a
+preliminary type for the type parameter `T`, so the type context used for
+inference of the function literal `(a, b) => a < b ? b : a` is `? Function(?,
+int)`.  Hence, `a` gets assigned a static type of `Object?`, and this leads to a
+compile error at `a < b`.
+
+With horizontal inference, type inference first visits the `0`, and uses its
+static type to assign a preliminary type of `int` to the type parameter `T`.
+Then, when it proceeds to perform type inference on the function literal, the
+type context is `int Function(int, int)`.  This results in `a` being assigned a
+static type of `int`, which is what the user intends.
+
+## Type inference algorithm for invocations
+
```

Comment:
Good points.  I've introduced a "terminology" section and I've tried to capture all the concepts I've used that are defined elsewhere.

Reply:
Thanks! It gets a bit heavy with all these explicit definitions, but I think it's very important to have the definitions somewhere. The language specification avoids being dominated by definitions only because it's longer, so we have the benefits of scale by defining a concept, say, on page 59, and then using it everywhere else in the document, relying on the index to find the definition if needed. So this document could be abbreviated somewhat by omitting some of the 4-5 line definitions, and just put a link to an existing definition in some other document at the first usage of the given word/phrase, but I don't think it's important to put a lot of time into that kind of improvement.

The main point is that it's absolutely great that we have definitions for a lot of things that were previously a bit harder to resolve!